src/main/java/org/squiddev/plethora/integration/astralsorcery/IntegrationAstralSorcery.java

```diff
@@ -0,0 +1,777 @@
+			if (constellation == null) return null;
+
+			ItemTunedCrystalBase.applyMainConstellation(crystalStack, constellation);
+
+			return crystalStack;
+		}
+	};
+
+	//Mainly useful if a player wants to design some sort of organizer to track their progress
+	public static final IMetaProvider<EntityPlayer> META_PLAYER_PROGRESS = new BaseMetaProvider<EntityPlayer>(
+		"Provides the player's progress in Astral Sorcery"
+	) {
+
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IPartialContext<EntityPlayer> context) {
+			EntityPlayer player = context.getTarget();
+			Map<String, Object> out = new HashMap<>();
+
+			PlayerProgress progress = ResearchManager.getProgress(player);
+
+			//Refers to the constellations that you have seen on a paper
+			out.put("seenConstellations", new LuaList<>(progress.getSeenConstellations()).asMap()); //FIXME ... great, it's the unlocalized name...
+
+			//Refers to the constellations that you have discovered via telescope, after seeing them on a paper
+			out.put("knownConstellations", new LuaList<>(progress.getKnownConstellations()).asMap()); //FIXME ... great, it's the unlocalized name...
+
+			out.put("availablePerkPoints", progress.getAvailablePerkPoints(player)); //SO MANY METHODS THAT SHOULD BE STATIC, NOT INSTANCE!!!
+
+			IConstellation attuned = progress.getAttunedConstellation();
+			out.put("attunedConstellation", attuned != null ? context.makePartialChild(attuned).getMeta() : null);
+			out.put("progressTier", progress.getTierReached().toString()); //REFINE Do we want the name, the ordinal, or a LuaList with both?
+
+			// ... shouldn't the `progressId` field be the same as the ordinal? ... whatever.
+			out.put("researchTier", progress.getResearchProgression().stream()
+				.max(Comparator.comparingInt(ResearchProgression::getProgressId))
+				.<Object>map(Enum::toString).orElse(null));
+
+			//REFINE Someone else can expose the Perks if they want; cost/benefit says "no" at this time
+
+			return Collections.singletonMap("astralProgress", out);
+		}
+
+		@Nonnull
+		@Override
+		public EntityPlayer getExample() {
+			return new EntityPlayerDummy(WorldDummy.INSTANCE);
+		}
+	};
+
+	public static final IMetaProvider<IConstellation> META_I_CONSTELLATION = new BasicMetaProvider<IConstellation>() {
+
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IConstellation context) {
+			Map<String, Object> out = new HashMap<>(7);
+
+			out.put("unlocalizedName", context.getUnlocalizedName());
+			out.put("simpleName", context.getSimpleName());
+
+			out.put("color", context.getConstellationColor().getRGB()); //Used for particles on rituals, collectors?
+			out.put("colour", context.getConstellationColor().getRGB());
+			out.put("tierColor", context.getTierRenderColor().getRGB());
+			out.put("tierColour", context.getTierRenderColor().getRGB());
+
+			//Because there isn't a nice, clean `getTier` method...
+			out.put("tier", context instanceof IMinorConstellation ? "minor"
```

Comment:
Can we shift this into a method or something? Just so we don't have nested ternaries. Sorry.

Reply:
That's fine; nested ternaries can be very messy.  I also put a `SuppressWarnings` annotation on the new method so it doesn't push nested ternaries.